libsolidity/parsing/Scanner.cpp

```diff
@@ -755,8 +768,21 @@ Token::Value Scanner::scanNumber(char _charSeen)
 				addLiteralCharAndAdvance();
 				if (!isHexDigit(m_char))
 					return Token::Illegal; // we must have at least one hex digit after 'x'/'X'
-				while (isHexDigit(m_char))
+				while (isHexDigit(m_char) || m_char == '_') // Unlike decimal digits, we keep the underscores for later validation
+				{
+					// Avoid double and trailing underscore
+					if (m_char == '_')
+					{
+						advance();
+						if (!isHexDigit(m_char))
+						{
+							rollback(1);
```

Comment:
I think control flow would be clearer if you move the `rollback(1)` before the if statement (and of course save `m_char` in a local variable).

Reply:
Ok, reading that again - why do you advance if you rollback later anyway?